tsdb/wal/wal_test.go

```diff
@@ -424,6 +425,94 @@ func TestCompression(t *testing.T) {
 	assert.Greater(t, float64(uncompressedSize)*0.75, float64(compressedSize), "Compressing zeroes should save at least 25%% space - uncompressedSize: %d, compressedSize: %d", uncompressedSize, compressedSize)
 }
 
+func TestLogPartialWrite(t *testing.T) {
+	const segmentSize = pageSize * 2
+	record := []byte{1, 2, 3, 4, 5}
+
+	tests := map[string]struct {
+		numRecords   int
+		faultyRecord int
+	}{
+		"partial write when logging first record in a page": {
+			numRecords:   10,
+			faultyRecord: 1,
+		},
+		"partial write when logging record in the middle of a page": {
+			numRecords:   10,
+			faultyRecord: 3,
+		},
+		"partial write when logging last record of a page": {
+			numRecords:   (pageSize / (recordHeaderSize + len(record))) + 10,
+			faultyRecord: pageSize / (recordHeaderSize + len(record)),
+		},
+		// TODO the current implementation suffers this:
+		//"partial write when logging a record overlapping two pages": {
+		//	numRecords:   (pageSize / (recordHeaderSize + len(record))) + 10,
+		//	faultyRecord: pageSize/(recordHeaderSize+len(record)) + 1,
+		//},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			dirPath, err := ioutil.TempDir("", "")
+			assert.NoError(t, err)
+
+			w, err := NewSize(nil, nil, dirPath, segmentSize, false)
+			assert.NoError(t, err)
+
+			// Replace the underlying segment file with a mocked one that injects a failure.
+			w.segment.SegmentFile = &faultySegmentFile{
+				SegmentFile:     w.segment.SegmentFile,
+				writeFailureAt:  ((recordHeaderSize + len(record)) * (testData.faultyRecord - 1)) + 1,
+				writeFailureErr: io.ErrShortWrite,
+			}
+
+			for i := 1; i <= testData.numRecords; i++ {
+				if err := w.Log(record); i == testData.faultyRecord {
+					assert.Error(t, io.ErrShortWrite, err)
+				} else {
+					assert.NoError(t, err)
+				}
+			}
+
+			assert.NoError(t, w.Close())
+
+			// Read it back. We expect no corruption.
+			s, err := OpenReadSegment(SegmentName(dirPath, 0))
+			assert.NoError(t, err)
+
+			r := NewReader(NewSegmentBufReader(s))
+			for i := 0; i < testData.numRecords; i++ {
+				assert.True(t, r.Next())
+				assert.NoError(t, r.Err())
+				assert.Equal(t, record, r.Record())
+			}
+		})
+	}
+}
+
+type faultySegmentFile struct {
+	SegmentFile
+
+	written         int
+	writeFailureAt  int
+	writeFailureErr error
+}
+
+func (f *faultySegmentFile) Write(p []byte) (int, error) {
+	if f.writeFailureAt <= f.written || f.writeFailureAt > f.written+len(p) {
+		n, err := f.SegmentFile.Write(p)
+		f.written += n
+		return n, err
+	}
+
+	// Inject failure.
+	partialLen := f.writeFailureAt - f.written
```

Comment:
Shall we add a check that `partialLen != len(p)` so that we make sure it's partial length (to avoid wrong use of this struct). 

We could change the above `if` condition to `f.writeFailureAt >= f.written+len(p)` because technically a full write should not return an error. Or better make it `partialLen--` if it is equal to `len(p)` while changing the definition of `writeFailureAt` that partial write is at that point or a byte before it (since it's a test, we can assume the record is >1 byte).

Reply:
I reversed the logic. Could you take a look, please? What do you think?